src/relax/backend/vm/rewrite_cuda_graph.cc

```diff
@@ -0,0 +1,410 @@
+    visited.clear();
+    std::unordered_set<const VarNode*> nodes_to_remove{src_nodes.begin(), src_nodes.end()};
+    auto f_sink = [&](const VarNode* node) {
+      if (visited.count(node)) return;
+      visited.insert(node);
+      auto it = graph->dep_src2dst.find(node);
+      if (it == graph->dep_src2dst.end()) return;
+      for (const auto* dst : it->second) {
+        nodes_to_remove.insert(dst);
+        task_stack.push_back(dst);
+      }
+    };
+    while (!task_stack.empty()) {
+      const VarNode* node = task_stack.back();
+      task_stack.pop_back();
+      f_sink(node);
+    }
+    // remove nodes from the graph
+    for (const auto* node : nodes_to_remove) {
+      graph->bindings.erase(node);
+    }
+    auto f_is_removed = [&](const VarNode* node) { return nodes_to_remove.count(node); };
+    graph->nodes.erase(std::remove_if(graph->nodes.begin(), graph->nodes.end(), f_is_removed),
+                       graph->nodes.end());
+    graph->alloc_tensors.erase(
+        std::remove_if(graph->alloc_tensors.begin(), graph->alloc_tensors.end(), f_is_removed),
+        graph->alloc_tensors.end());
+  }
+
+  Function EmitStaticAllocations(BlockBuilderNode* builder, StaticRegion* region) {
+    builder->BeginBindingBlock();
+    Array<Expr> alloc_tensors;
+    Array<StructInfo> alloc_tensors_struct_info;
+    for (const auto& var : scope_.graph.alloc_tensors) {
+      region->alloc_tensor_to_index.emplace_back(var, alloc_tensors.size());
+      alloc_tensors.push_back(builder->Emit(scope_.graph.bindings.at(var)->value));
+      alloc_tensors_struct_info.push_back(
+          Downcast<StructInfo>(GetRef<Var>(var)->struct_info_.value()));
+    }
+    auto output = builder->Emit(Tuple(alloc_tensors));
+    auto block = builder->EndBlock();
+    auto func_body = builder->Normalize(SeqExpr({block}, output));
+    auto func = Function({}, func_body, TupleStructInfo(alloc_tensors_struct_info));
+    return func;
+  }
+
+  Function EmitStaticGraph(BlockBuilderNode* builder, StaticRegion* region) {
+    Var allocs{"allocs", region->alloc_func->ret_struct_info};
+    builder->BeginBindingBlock();
+    for (int i = 0; i < static_cast<int>(scope_.graph.alloc_tensors.size()); ++i) {
+      const VarNode* tensor = scope_.graph.alloc_tensors[i];
+      const Expr alloc = builder->Normalize(TupleGetItem(allocs, i));
+      builder->EmitNormalized(VarBinding(GetRef<Var>(tensor), alloc));
+    }
+    for (const VarNode* node : scope_.graph.nodes) {
+      const VarBindingNode* binding = scope_.graph.bindings.at(node);
+      builder->EmitNormalized(GetRef<VarBinding>(binding));
+    }
+    auto void_output = builder->Emit(Tuple(Array<Expr>{}));
+    auto block = builder->EndBlock();
+    auto func_body = builder->Normalize(SeqExpr({block}, void_output));
+    auto func = Function({allocs}, func_body, TupleStructInfo(Array<StructInfo>{}));
+    return func;
+  }
+
+  void SummarizeStaticRegion() {
+    PruneOutputNodes(&scope_.graph, outputs_);
```

Comment:
Why is it necessary to exclude calls related to outputs? It seems the allocation and address can be statically initialized by the vm.

Reply:
The output tensor is visible to users. If we don't prune the output, one implication is that when the model is run the second time, the output tensor of the first run get overwritten.

cc @tqchen

Reply:
I agree, we can have another mode where output is allocated, but for most general mode, we should skip input and outputs